### PR TITLE
feat: Add `CostTrigger`

### PR DESCRIPTION
**Pull request by robinholzi**
# Motivation

Following up https://github.com/eth-easl/modyn/pull/599, https://github.com/eth-easl/modyn/pull/602, https://github.com/eth-easl/modyn/pull/604 we want to implement a CostTrigger base class that reflects the tradeoff-based trigger between a cost and a regret metric.

Furthermore, we add a data incorporation latency-based cost trigger as the first variant (subclass) of `CostTrigger`.

https://github.com/eth-easl/modyn/pull/605 will add a second subclass based on misclassifications

### CODE REVIEW

File: modyn/supervisor/internal/triggers/cost/avoidablemissclassification_costtrigger.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    ) -> None:
+        """Update the cost and performance trackers with the new model
+        metadata."""
+        super().inform_new_model(most_recent_model_id, number_samples, training_time)
+
+    # ---------------------------------------------------------------------------------------------------------------- #
+    #                                                     INTERNAL                                                     #
+    # ---------------------------------------------------------------------------------------------------------------- #
+
+    @override
```
Comment (robinholzi) [inline modyn/supervisor/internal/triggers/cost/avoidablemissclassification_costtrigger.py:45]: Will be addressed in a followup PR once `PerformanceTrigger` is fully integrated

File: modyn/supervisor/internal/triggers/cost/costtrigger.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    def init_trigger(self, context: TriggerContext) -> None:
+        self.context = context
+
+    def inform(
+        self,
+        new_data: list[tuple[int, int, int]],
+        log: TriggerPolicyEvaluationLog | None = None,
+    ) -> Generator[int, None, None]:
+        new_key_ts = self._leftover_data + [(key, timestamp) for key, timestamp, _ in new_data]
+        # reappending the leftover data to the new data requires incrementing the sample left until detection
```
Comment (robinholzi) [inline modyn/supervisor/internal/triggers/cost/costtrigger.py:53]: These 20 lines are duplicated across all policies that use discretized time units for detection. I think this is okay. An abstraction for this with polymorphic delegation functions would make the code quite uncomprehensive, as every trigger has some special data structures that need to be updated within this section.

Comment (robinholzi) [reply modyn/supervisor/internal/triggers/cost/costtrigger.py:54]: ** Resolved via BatchedTrigger

File: modyn/supervisor/internal/triggers/cost/avoidablemissclassification_costtrigger.py (hunk 1 above)
Comment (robinholzi) [reply modyn/supervisor/internal/triggers/cost/avoidablemissclassification_costtrigger.py:47]: https://github.com/eth-easl/modyn/pull/605

File: modyn/supervisor/internal/triggers/cost/avoidablemissclassification_costtrigger.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
+from __future__ import annotations
```
Comment (robinholzi) [inline modyn/supervisor/internal/triggers/cost/avoidablemissclassification_costtrigger.py:1]: will be implemented and tested in https://github.com/eth-easl/modyn/pull/605

**Review by MaxiBoether — APPROVED** [review]
actually i just read through the costtrigger and latency trigger and due to the nice abstractions, at least I could immediately understand what's happening :) so LGTM!

**Pull request merged**